### PR TITLE
feat(pydantic): allow pydantic model instances as defaults (#64)

### LINKED ISSUE #64: pydantic classes as default values not serializable
Hi everyone,

I have a pydantic model that has an attribute of another pydantic model type that should be generated by default. This fails as the schema dict produced by py_avro_schema sets as default the pydantic class object, which is per default not json serializable via orjson.

example.py

```
import py_avro_schema as pas
from pydantic import BaseModel, Field


class Bar(BaseModel):
    baz: int = 0


class Foo(BaseModel):
    bar: Bar = Field(default_factory=Bar)


pas.generate(Foo)
```

This raises the following error:

```
Traceback (most recent call last):
  File "/Users/gellertd/workspace/procureai/foundation/constellation/example.py", line 13, in <module>
    pas.generate(Foo)
  File "/Users/gellertd/.pyenv/versions/constellation/lib/python3.11/site-packages/memoization/caching/plain_cache.py", line 42, in wrapper
    result = user_function(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gellertd/.pyenv/versions/constellation/lib/python3.11/site-packages/py_avro_schema/__init__.py", line 69, in generate
    schema_json = orjson.dumps(schema_dict, option=json_options)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Type is not JSON serializable: Bar
```

**Suggestion**

On a PydanticSchema for each recordfield check if the `py_type` is a pydantic basemodel subclass and if so call
`default.model_dump(mode="json")` on it

This would then produce this schema:

```
{
  "type": "record",
  "name": "Foo",
  "fields": [
    {
      "name": "bar",
      "type": {
        "type": "record",
        "name": "Bar",
        "fields": [{ "name": "baz", "type": "long", "default": 0 }],
        "namespace": "__main__",
        "doc": "Usage docs: https://docs.pydantic.dev/2.6/concepts/models/"
      },
      "default": { "baz": 0 }
    }
  ],
  "namespace": "__main__",
  "doc": "Usage docs: https://docs.pydantic.dev/2.6/concepts/models/"
}
```

which looks alright.

The downside is that pydantic would be needed at runtime so it must be either catched as an error or it would be imported on class / method level.

Let me know what you think about this please.
Thanks a lot.

### PR DESCRIPTION
**Pull request by dada-engineer**
This allows pydantic BaseModels as default values.

Fixes #64

### CODE REVIEW

**Comment by dada-engineer** [comment]
linting fails because black and flake8 have a different opinion on how

```
class PyType(str):
    ...
```

should be formatted.

black now wants to do:

```
class PyType(str): ...
```

But flake8 then complains that there are multiple statements on one line.

File: src/py_avro_schema/_schemas.py [hunk 1]
```diff
@@ -901,6 +905,10 @@ def _record_field(self, name: str, py_field: pydantic.fields.FieldInfo) -> Recor
         )
         return field_obj
 
+    def make_default(self, py_default: Any) -> Any:
+        """Return an Avro schema compliant default value for a given Python value"""
+        return {key: _schema_obj(value.__class__).make_default(value) for key, value in py_default}
```
Comment (dada-engineer) [inline src/py_avro_schema/_schemas.py:910]: @faph can we assume it is a pydantic class here so it is iterable like this?

Comment (faph) [reply src/py_avro_schema/_schemas.py:910]: @dada-engineer Correct, and the type hint should reflect this.

Also, return type should be `JSONObj`

**Comment by faph** [comment]
So this flake8 vs black conflict. Is there a way that we can configure either of the tools to handle this OK?

**Comment by dada-engineer** [comment]
> So this flake8 vs black conflict. Is there a way that we can configure either of the tools to handle this OK?

I guess this comment makes sense: https://github.com/psf/black/issues/4213#issuecomment-1928364118

TLDR;
deactivate flake8 rule for stuff that is handled by formatter.

File: src/py_avro_schema/_schemas.py (hunk 1 above)
Comment (dada-engineer) [reply src/py_avro_schema/_schemas.py:907]: done

**Comment by faph** [comment]
Got you, so that should be a change in `.flake8` file?

**Comment by dada-engineer** [comment]
Looks like it, should I add this here?

**Comment by dada-engineer** [comment]
Had some issue with mypy as well, fixed in same commit.

**Comment by faph** [comment]
Love it! 

Are we good to merge?

**Comment by dada-engineer** [comment]
I am okay with it, but since I removed it lists of pydantic objects will not yet work. I'll open a new issue for this or a second PR later.

**Comment by faph** [comment]
Understood. Many thanks for this contribution.

**Comment by koenlek** [comment]
Thanks for making this fix! Unfortunately, I'm stuck on py-avro-schema v2 (and upgrading is going to be very hard I'm afraid). Any suggestions on how I could try and patch v2 or work around this issue?

I tried patching the make_default logic into v2, but it doesn't seem to work unfortunately...

Any hints would be highly appreciated 🙏

**Comment by koenlek** [comment]
NVM. Gotten it to work and posted it for others here: https://github.com/jpmorganchase/py-avro-schema/issues/64#issuecomment-2388957280
